### ISSUE TITLE
Consider reducing MaxTableSize

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ds-badger#17]
It's currently set to 64MiB. Unfortunately, this means that badger will allocate 82MiB of memory up-front and never let go of it.

My worry is that this probably can't be changed after initializing the datastore. Is that the case? It would be *nice* to be able to tune this parameter in the config.

**Comment by schomatis** [comment]
> My worry is that this probably can't be changed after initializing the datastore. Is that the case?

I think you're right. The options ([`DB.opt`](https://github.com/dgraph-io/badger/blob/2e3a32f0ccac3066fb4206b28deb39c210c5266f/db.go#L253)) are not exported and only written at initialization (when the DB is opened/created) and I don't think they would want to export that as many of the options we're not designed to be changed at runtime. Maybe our datastore could be enhanced to reopen the DB with a different `MaxTableSize` value if memory consumption is an issue but I'm not sure if the datastore service can be interrupted without disturbing other services.

**Comment by Stebalien** [comment]
Doing this at runtime would be awesome but I'm not sure it's even possible to change this after creating the datastore (even if we restart). That is, I'd like to make this a config option that can be tweaked after the datastore has been initialized but I'm not sure if that's possible (without rebuilding the datastore).

**Comment by schomatis** [comment]
> but I'm not sure it's even possible to change this after creating the datastore (even if we restart).

On the Badger side it is, I'm not sure if the options were designed to be modified *after* creation of the DB, but specifically `MaxTableSize` can be modified because it basically controls the size of the [arena](https://github.com/dgraph-io/badger/blob/master/skl/arena.go#L48) (a byte buffer) for the in-memory table that is created and destroyed (flushed to disk) every time the DB is opened/created (or when the table is filled and flushed to disk).

On the datastore side I don't see a problem as its basically a wrapper for the Badger DB.

**Comment by Stebalien** [comment]
I see. So shrinking this after the fact will cause new tables to be smaller but won't actually break anything (as far as I can tell). I was worried that there was some code that assumed that the size of an on-disk table is *always* less than MaxTableSize.

**Comment by schomatis** [comment]
> So shrinking this after the fact will cause new tables to be smaller but won't actually break anything (as far as I can tell).

No (also as far as I can tell).

**Comment by manishrjain** [comment]
I think the biggest consumer are two:

1. mmap-ing value log. That could be put purely on disk.
2. Tables in RAM. They could be mmapped.

I'd also suggest looking into the new LSMOnly option:
https://github.com/dgraph-io/badger/blob/be3a85f6a32751997ddee1b298c2113aa7bec157/options.go#L136

This still uses value log but tries to cram more values into the LSM tree, hence you'd be able to reclaim the disk space quicker (if disk space is a concern -- and if value log is mmapped, then that'd translate in RAM savings too).

**Comment by Stebalien** [comment]
I've reduced the max size to 16MiB and haven't noticed any issues. I've also switched to FileIO for the value log and saw no performance loss on write and a 4x performance increase on read...

Note: many of the values in my tests are 256KiB (i.e., _HUGE_). That's probably why I saw the perf increase.

Also note: inlining 16KiB values probably isn't going to help us too much in this particular case. But we can keep that in mind.
